passes/memory/memory_map.cc

```diff
@@ -356,17 +407,71 @@ struct MemoryMapPass : public Pass {
 	{
 		//   |---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|
 		log("\n");
-		log("    memory_map [selection]\n");
+		log("    memory_map [options] [selection]\n");
 		log("\n");
 		log("This pass converts multiport memory cells as generated by the memory_collect\n");
 		log("pass to word-wide DFFs and address decoders.\n");
 		log("\n");
+		log("    -attr !<name>\n");
+		log("        do not map memories that have attribute <name> set.\n");
+		log("\n");
+		log("    -attr <name>[=<value>]\n");
+		log("        for memories that have attribute <name> set, only map them if its value\n");
+		log("        is a string <value> (if specified), or an integer 1 (otherwise). if this\n");
+		log("        option is specified multiple times, map the memory if the attribute is\n");
+		log("        to any of the values.\n");
+		log("\n");
+		log("    -iattr\n");
+		log("        for -attr, ignore case of <value>.\n");
+		log("\n");
 	}
-	void execute(std::vector<std::string> args, RTLIL::Design *design) YS_OVERRIDE {
+	void execute(std::vector<std::string> args, RTLIL::Design *design) YS_OVERRIDE
+	{
+		bool attr_icase = false;
+		dict<RTLIL::IdString, std::vector<RTLIL::Const>> attributes;
+
 		log_header(design, "Executing MEMORY_MAP pass (converting $mem cells to logic and flip-flops).\n");
-		extra_args(args, 1, design);
-		for (auto mod : design->selected_modules())
-			MemoryMapWorker(design, mod);
+
+		size_t argidx;
+		for (argidx = 1; argidx < args.size(); argidx++)
+		{
+			if (args[argidx] == "-attr" && argidx + 1 < args.size())
+			{
+				std::string attr_arg = args[++argidx];
+				std::string name;
+				RTLIL::Const value;
+				size_t eq_at = attr_arg.find('=');
+				if (eq_at != std::string::npos) {
+					name  = attr_arg.substr(0, eq_at);
+					value = attr_arg.substr(eq_at + 1);
+				} else {
+					name  = attr_arg;
+					value = RTLIL::Const(1);
+				}
+				if (attr_arg.size() > 1 && attr_arg[0] == '!') {
+					if (value != RTLIL::Const(1))
+						break; // we don't support -attr !<name>=<value>
```

Comment:
At this point we have already executed `++argidx`, therefore it will look like `extra_args()` as if `!<name>=<value>` is a select expression. We need to `--argidx` before break to make sure `extra_args()` sees `-attr` and produces an `Unknown option or option in arguments.` error message. Otherwise the changes to memory_map LGTM.